src/Sylius/Component/Order/Model/OrderItem.php

```diff
@@ -131,103 +124,133 @@ public function setUnitPrice($unitPrice)
-            $adjustment->setAdjustable(null);
-            $this->adjustments->removeElement($adjustment);
+        $this->unitsTotal = 0;
+
+        foreach ($this->units as $unit) {
+            $this->unitsTotal += $unit->getTotal();
         }
+
+        $this->recalculateTotal();
     }
 
     /**
      * {@inheritdoc}
      */
-    public function hasAdjustment(AdjustmentInterface $adjustment)
+    public function equals(OrderItemInterface $orderItem)
     {
-        return $this->adjustments->contains($adjustment);
+        return $this === $orderItem;
     }
 
     /**
      * {@inheritdoc}
      */
-    public function getAdjustmentsTotal($type = null)
+    public function isImmutable()
     {
-        if (null === $type) {
-            return $this->adjustmentsTotal;
-        }
-
-        $total = 0;
-        foreach ($this->getAdjustments($type) as $adjustment) {
-            $total += $adjustment->getAmount();
-        }
-
-        return $total;
+        return $this->immutable;
     }
 
     /**
      * {@inheritdoc}
      */
-    public function removeAdjustments($type)
+    public function setImmutable($immutable)
     {
-        foreach ($this->getAdjustments($type) as $adjustment) {
-            if ($adjustment->isLocked()) {
-                continue;
-            }
+        $this->immutable = (bool) $immutable;
+    }
 
-            $adjustment->setAdjustable(null);
-            $this->adjustments->removeElement($adjustment);
-        }
+    /**
+     * {@inheritdoc}
+     */
+    public function getUnits()
+    {
+        return $this->units;
     }
 
     /**
      * {@inheritdoc}
      */
-    public function clearAdjustments()
+    public function addUnit(OrderItemUnitInterface $unit)
     {
-        return $this->adjustments->clear();
+        if ($this !== $unit->getOrderItem()) {
+            throw new \LogicException('This order item unit is assigned to a different order item.');
+        }
+
+        if (!$this->hasUnit($unit)) {
+            $this->units->add($unit);
+
+            $this->quantity++;
+            $this->unitsTotal += $unit->getTotal();
+            $this->recalculateTotal();
+
+            if (null !== $this->order) {
+                $this->order->calculateItemsTotal();
+            }
+        }
     }
 
     /**
      * {@inheritdoc}
      */
-    public function calculateAdjustmentsTotal()
+    public function removeUnit(OrderItemUnitInterface $unit)
     {
-        $this->adjustmentsTotal = 0;
+        if ($this->hasUnit($unit)) {
+            $this->units->removeElement($unit);
 
-        foreach ($this->adjustments as $adjustment) {
-            if (!$adjustment->isNeutral()) {
-                $this->adjustmentsTotal += $adjustment->getAmount();
+            $this->quantity--;
+            $this->unitsTotal -= $unit->getTotal();
+            $this->recalculateTotal();
+
+            if (null !== $this->order) {
+                $this->order->calculateItemsTotal();
```

Comment:
I would say this should be called each time `$this->recalculateTotal()` method is called, so we can move it there.